exercise-book/src/iterators.md

```diff
@@ -0,0 +1,324 @@
+- collect a result to different containers
+- turn off rust-analyzer inlay hints
+
+## Prerequisites
+
+For completing this exercise you need to have
+
+- knowledge of control flow
+- how to write basic functions
+- know basic Rust types
+
+## Task
+
+- Add the odd numbers in the following string using an iterator chain
+
+```text
+//ignore everything that is not a number
+1
+2
+3
+4
+five
+6
+7
+∞
+9
+X
+```
+
+- Take the template in [exercise-templates/iterators](../../exercise-templates/iterators/) as a starting point.
+- Replace the first `todo!` item with [reader.lines()]() and continue "chaining" the iterators until you've calculated the desired result.
+- Run the code with `cargo run --bin iterators1` when inside the `exercise-templates` directory
+
+If you need it, we have provided a [complete solution](../../exercise-solutions/iterators/src/bin/iterators1.rs) for this exercise.
+
+## Knowledge
+
+### Iterators and iterator chains
+
+Iterators are a way to write for loops in a functional style. The main idea is to take away the error prone indexing and control flow by giving them a name that you and others can understand and compose safely.
+
+For example, to double every number given by a vector, you could write a for loop:
+
+```rust
+let v = [10, 20, 30];
+let mut xs = [0, 0, 0];
+
+for idx in 0..=v.len() {
+  xs[idx] = 2 * v[idx];
+}
+```
+
+In this case, the name we give to the procedure `2 * v[idx]` and juggling the index over the entire collection is a [map](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.map). An idiomatic Rustacean would write something similar to the following (period indented) code.
+
+```rust
+let v = [10, 20, 30];
+let xs: Vec<_> = v.iter()
+                  .map(|elem| elem * 2)
+                  .collect();
+```
+
+No win for brevity, but it has several benefits:
+
+- Changing the underlying logic is more robust
+- Less indexing operations means you will fight the borrow checker less in the long run
+- You can parallelize your code with minimal changes using [rayon](https://crates.io/crates/rayon).
+
+The first point is not in vain - the original snippet has a bug in the upper bound, since `0..=v.len()` is inclusive!
+
+Think of iterators as lazy functions - they only carry out computation when called with a `.collect()` or similar, not the `.map()` itself.
+
+### Turbo fish syntax `::<>`
+
+Iterators sometimes struggle to figure out the types of all intermediate steps and need assistance.
+
+We can write
+
+```rust
+let xs = v.iter()
+          .map(|elem| elem * 2)
+          .collect::<Vec<i32>>();
+```
+
+instead to avoid having a `xs: Vec<_> = ...`. This `::<SomeType>` syntax is called the [turbo fish operator](https://doc.rust-lang.org/book/appendix-02-operators.html?highlight=turbo%20fish#non-operator-symbols), and it disambiguates calling the same method with different output types, like `.collect::<HashSet<i32>>()` and `.collect::<Vec<i32>>()` (try it!)
```

Comment:
>  to avoid having a `xs: Vec<_>

while it is true, many codebases prefer / mandate the use of a type hint on the last over turbofish, which many people find less readable. Rust Analyzer does it, for example, and many other Rust projects follow suit. So, an example when turbo fish is used in `collect` may not sell the idea well. I think a better example would be a turbo fish inside an iterator closure:

```rust
let numbers: Vec<_> = ["1", "2", "3"]
    .iter()
    .map(|s| s.parse::<i32>().unwrap())
    // a turbofish in `parse` call above
    // helps a compiler determine the type of `n` below
    .map(|n| n + 1)
    .collect();
```